File: TBB/src/research/Tree.cpp
#include "Tree.h"

// - Note
//      No side effect in the function. 
//      Explicit assignment is required after calculation
// - Example
//      auto tuple = Calculate(tree, R, C);
//      tree.root[R][C] = tuple.root;
//      tree.cost[R][C] = tuple.cost;
  auto Tree::Calculate(const Tree & tree, i32 _row, i32 _col) noexcept
      -> std::tuple<i64, f64>
{
    i64  best_root = -1;
    f64  weight{}, best_weight = LDBL_MAX;

    // Unused range
    if (_row >= _col) {
        weight = 0;
    }
    // Main diagonal
    else if (_row + 1 == _col) {
        best_root = _row + 1;
        weight = tree.prob[_row];
    }
    // Tree estimation
    else {
        f64 sum = 0; // basic weight
        for (i32 i = _row; i < _col; ++i)
        {
            // Accumulate
            sum += tree.prob[i];

            // Find best weight
            f64 temp_weight =
                tree.cost[_row][i] + tree.cost[i + 1][_col];

            if (temp_weight < best_weight) {
                best_weight = temp_weight;
                best_root = i + 1;
            }
        }
        weight = best_weight + sum;
    }
    // Weight == Cost
    return std::make_tuple(best_root, weight);
}

// - Note
//      Number of vertices in tree

  size_t Tree::size() const noexcept
{
    return prob.size();
}

// - Note
//      Initialize tree's probabilities with random value
void Init(Tree & _tree)
{
    using namespace std;
    using namespace std::chrono;

    // Use epoch time as the seed of random generator
    auto duration = system_clock::now().time_since_epoch();
    auto seed = static_cast<u32>(duration.count());
    mt19937 rnd{ seed };

    f64     total{};
    for (i32 i = 0; i < _tree.prob.size(); ++i) {
        total += _tree.prob[i] = rnd() % (1 << 20);
    }
    // Normalization
    for (i32 i = 0; i < _tree.prob.size(); ++i) {
        _tree.prob[i] /= total;
    }
}

// - Note
//      Display the tree in console. 
//      Used for debugging
void Display(const Tree & _tree)
{
    using namespace std;

    const i32 N = static_cast<i32>(_tree.size());

    puts("--- Prob ---");
    for (i32 i = 0; i < N; ++i) {
        printf_s(" %4.4lf, ", _tree.prob[i]);
    }
    putchar('\n');

    puts("--- Root/Cost ---");
    for (i32 i = 0; i < N + 1; ++i) {
        for (i32 j = 0; j < N + 1; ++j)
        {
            auto root = _tree.root[i][j];
            auto cost = _tree.cost[i][j];
            printf_s(" [%2lld,%2.2lf] ", root, cost);
        }
        putchar('\n');
    }
}

bool operator==(const Tree& lhs, const Tree& rhs) noexcept
{
    if (lhs.prob != rhs.prob) {
        return false;
    }
    const i32 N = static_cast<i32>(lhs.size());

    // Compare valid range(x) only.
    //   [ - x x x ]
    //   [ - - x x ]
    //   [ - - - x ]
    //   [ - - - - ]
    for (i32 i = 0; i < N; ++i) {
        for (i32 j = i + 1; j < N; ++j)
        {
            const bool equal_root = lhs.root[i][j] != rhs.root[i][j];
            const bool equal_cost = lhs.cost[i][j] != rhs.cost[i][j];

            if (equal_root == false || equal_cost == false) {
                return false;
            }
        }
    }
    return true;
}

File: golang/src/tree.go
// ==== ==== ==== ==== ==== ==== ==== ==== ==== ==== ==== ==== ====
//
//  Author 	: Park  Dong Ha ( luncliff@gmail.com )
//
// 	Note	:
//      Optimal Binary Search Tree for Dynamic Programming
//
//  See also :
//      `package matrix`
//  Reference :
//      https://www.cs.auckland.ac.nz/software/AlgAnim/opt_bin.html
//
// ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ----
package main

import (
	"fmt"
	"io"
	"math"
	"math/rand"
	"matrix"
)

// Tree ...
//  	Custom structure for Optimal Binary Search Tree
//  	- `prob` : []float64
// 		- `root` : [][]int
//		- `cost` : [][]float64
type Tree struct {
	Prob []float64
	Cost matrix.Float2D // Float64 2D matrix
	Root matrix.Int2D   // Int 2D matrix
}

// NewTree ...
//		Allocate memory resources for Optimal Binary Search Tree
//  	See also : type `Tree`
func NewTree(n int) *Tree {
	res := new(Tree)
	res.Prob = make([]float64, n)
	res.Root = matrix.MakeInt2D(n+1, n+1)
	res.Cost = matrix.MakeFloat2D(n+1, n+1)

	return res
}

// Init ...
//  	Setup probabilities of the tree's vertices
func (tree *Tree) Init() {
	var total float64
	// Distribute random values
	for i := range tree.Prob {
		R := rand.Float64()
		total += R
		tree.Prob[i] = R
	}
	// Normalize with total value
	for i := range tree.Prob {
		tree.Prob[i] = tree.Prob[i] / total
	}
}

// Size ...
//  	The number of vertices in tree
func (tree *Tree) Size() int {
	return len(tree.Prob)
}

// Calculate ...
//  	To prevent side effect,
//		explicit assignment is required after calculation.
// 		Tree won't be modified in the function
func (tree *Tree) Calculate(row int, col int) (root int, weight float64) {
	var bestWeight float64 = math.MaxFloat64 // Optimal Cost
	var bestRoot int = -1                    // Optimal Root

	switch {
	// Unused range
	case row >= col:
		root, weight = -1, 0.0

	// Main diagonal
	case row+1 == col:
		root, weight = row+1, tree.Prob[row]

	// Estimation : (Data Dependency Exists)
	case row+1 < col:
		sum := 0.0 // basic weight of tree

		for i := row; i < col; i++ {
			sum += tree.Prob[i] // Accumulate

			// Find optimized case
			tempWeight := *tree.Cost.At(row, i) + *tree.Cost.At(i+1, col)
			if tempWeight < bestWeight {
				bestWeight = tempWeight
				bestRoot = i + 1 // 1-based indexing
			}
		}
		// Root : Optimal root index
		// Cost : Sum of all cost + Additinal weight
		root, weight = bestRoot, bestWeight+sum
	}

	return // return tuple
}

func (lhs *Tree) Copy(rhs *Tree) {
	N := rhs.Size()

	for i := 0; i < N; i++ {
		lhs.Prob[i] = rhs.Prob[i]
	}
	for i := 0; i <= N; i++ {
		for j := 0; j <= N; j++ {
			*lhs.Root.At(i, j) = *rhs.Root.At(i, j)
			*lhs.Cost.At(i, j) = *rhs.Cost.At(i, j)
		}
	}
}

// Equal...
//  	Check if both tree is equal. Only consider valid indexes
func (lhs *Tree) Equal(rhs *Tree) bool {
	// Equal Size?
	if lhs.Size() != rhs.Size() {
		return false
	}
	N := lhs.Size()

	// Equal Probability ?
	for i := 0; i < N; i++ {
		if lhs.Prob[i] != rhs.Prob[i] {
			// log.Println("Different Prob")
			return false
		}
	}

	// Equal Root & Cost ?
	for i := 0; i < N; i++ {
		for j := i + 1; j < N; j++ {
			equalroot := *lhs.Root.At(i, j) != *rhs.Root.At(i, j)
			equalcost := *lhs.Cost.At(i, j) != *rhs.Cost.At(i, j)
			if equalroot == false || equalcost == false {
				return false
			}
		}
	}
	return true
}

func (tree *Tree) Display(out io.Writer) {
	N := tree.Size()

	for i := 0; i <= N; i++ {
		for j := 0; j <= N; j++ {
			fmt.Fprintf(out, " [%2d, %2.2f]",
				*tree.Root.At(i, j), *tree.Cost.At(i, j))
		}
		fmt.Fprintln(out)
	}
}
